cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/submit/AbstractSparkSubmitter.java

```diff
@@ -216,14 +218,61 @@ protected void submit(SparkRuntimeContext runtimeContext,
 
   private abstract static class SparkJobFuture<V> extends AbstractFuture<V> {
 
+    private static final Logger LOG = LoggerFactory.getLogger(SparkJobFuture.class);
+    private final AtomicBoolean done;
+    private final SparkRuntimeContext context;
+
+    protected SparkJobFuture(SparkRuntimeContext context) {
+      this.done = new AtomicBoolean();
+      this.context = context;
+    }
+
     @Override
-    public boolean set(@Nullable V value) {
-      return super.set(value);
+    protected boolean set(V value) {
+      if (done.compareAndSet(false, true)) {
+        return super.set(value);
```

Comment:
what if super.set() returns false? That means it was already set, but not through this method?